source/Crew.cpp

```diff
@@ -0,0 +1,223 @@
+	isPaidSalaryWhileParked = false;
+	dailySalary = 100;
+	minimumPerShip = 0;
+	populationPerOccurrence = 0;
+	
+	for(const DataNode &child : node)
+	{
+		if(child.Size() >= 2)
+		{
+			if(child.Token(0) == "on escorts?")
+				isOnEscorts = child.Value(1);
+			else if(child.Token(0) == "on flagship?")
+				isOnFlagship = child.Value(1);
+			else if(child.Token(0) == "pay salary while parked?")
+				isPaidSalaryWhileParked = child.Value(1);
+			else if(child.Token(0) == "daily salary")
+				dailySalary = child.Value(1);
+			else if(child.Token(0) == "minimum per ship")
+				minimumPerShip = child.Value(1);
+			else if(child.Token(0) == "population per occurrence")
+				populationPerOccurrence = child.Value(1);
+			else if(child.Token(0) == "name")
+				name = child.Token(1);
+			else
+				child.PrintTrace("Skipping unrecognized attribute:");
+		}
+		else
+			child.PrintTrace("Skipping incomplete attribute:");
+	}
+}
+
+
+
+int64_t Crew::CalculateSalaries(
+	const vector<shared_ptr<Ship>> ships,
+	const bool includeExtras
+) {
+	int64_t totalSalaries = 0;
+
+	for(const shared_ptr<Ship> &ship : ships)
+	{
+		totalSalaries += Crew::SalariesForShip(
+			ship,
+			// Determine whether if the current ship is the flagship
+			ship->Name() == ships.front()->Name(),
+			includeExtras
+		);
+	}
+
+	return totalSalaries;
+}
+
+
+
+int64_t Crew::CostOfExtraCrew(const vector<shared_ptr<Ship>> ships)
+{
+	// Calculate the salaries with and without extras and return the difference.
+	return Crew::CalculateSalaries(ships, true)
+		- Crew::CalculateSalaries(ships, false);
+}
+
+
+
+int64_t Crew::NumberOnShip(
+	const Crew crew,
+	const shared_ptr<Ship> ship,
+	const bool isFlagship,
+	const bool includeExtras
+) {
+	int64_t count = 0;
+
+	// 
+	// If this is the flagship, check if this kind of crew appears on the flagship.
+	if(isFlagship && !crew.IsOnFlagship()) return count;
+	// If this is an escort, check if this kind of crew appears on escorts.
+	if(!isFlagship && !crew.IsOnEscorts()) return count;
+	
+	const int64_t countableCrewMembers = includeExtras
+		? ship->Crew()
+		: ship->RequiredCrew();
+
+	// Apply the per-ship minimum.
+	 count = min(crew.MinimumPerShip(), countableCrewMembers);
+
+	// Guard against division by zero to prevent the universe from imploding.
+	if(crew.PopulationPerOccurrence())
+	{
+		// Figure out how many of this kind of crew we have, by population.
+		count = max(count, countableCrewMembers / crew.PopulationPerOccurrence());
+	}
+
+	return count;
+}
+
+
+
+int64_t Crew::SalariesForShip(
+	const shared_ptr<Ship> ship,
+	const bool isFlagship,
+	const bool includeExtras
+) {
+	// We don't need to pay dead people.
+	if(ship->IsDestroyed()) return 0;
+	
+	const Set<Crew> crews = GameData::Crews();
+	const Crew *defaultCrew = crews.Find("default");
+
+	int64_t salariesForShip = 0;
+	int64_t specialCrewMembers = 0;
+
+	// Add up the salaries for all of the special crew members
+	for(const pair<const string, Crew> crewPair : crews)
```

Comment:
XCode gives me a warning here:

> Loop variable 'crewPair' of type 'const pair<const std::__1::string, Crew>' (aka 'const pair<const basic_string<char, char_traits<char>, allocator<char> >, Crew>') creates a copy from type 'const pair<const std::__1::string, Crew>'

It suggests the following fix:

> 1. Use reference type 'const pair<const std::__1::string, Crew> &' (aka 'const pair<const basic_string<char, char_traits<char>, allocator<char> >, Crew> &') to prevent copying

Implementing the suggestion completely breaks the feature, though. With the change, the game doesn't count any non-default crew members.

Any ideas?